File: book/check_refusals.R
# check_refusals.R — does every documented refusal actually refuse?
#
# Run from the repository root; sourced by the R test suite.
#
# **`#| error: true` does not assert that a chunk errors. It tolerates one.** So
# a chunk the book presents as a refusal can quietly stop refusing, render a
# perfectly good table, and every build still exits 0. The prose above it goes on
# claiming the grammar catches something it no longer catches, and no guard in
# this directory can see it: `check_grammar.R` asks whether a pipeline parses,
# and this kind of chunk is one that is *supposed* not to.
#
# It is one of the three defects the writing guide names as uncatchable by
# anything else, and the sibling project shipped exactly this: one of its 48
# refusal chunks drew an empty picture instead of refusing, for several sessions.
#
# **A pipeline is lazy here, which is the trap specific to god.** Evaluating
# `sales |> keep(nonsense)` returns a pipeline object and raises nothing at all.
# It is `collect`, or printing it, that hands the sentence to the engine. So a
# check that only evaluated the chunk would report every refusal as passing while
# testing none of them. The forcing below is the whole reason this file is not
# three lines long.
#
# **An error is not automatically a refusal, either.** A chunk that failed
# because this file did not hand it a table looks identical to one that failed
# because the grammar said no, and that turns a broken check into a green one.
# Which errors mean *this file* is broken is the short, stable list, so that is
# the one written down below rather than a list of what refusals look like.

check_refusals <- function(dirs = "book") {
  # A chapter's chunks are evaluated below for their side effects, and one
  # side effect a guard must never keep is opening windows. A gog plot
  # printed outside a notebook writes a temp page and hands it to the
  # system browser, so the chapter that draws with the sibling opened
  # seven tabs on every run of this file; the render itself never does,
  # because knitr embeds the plot instead of printing it. The browser is a
  # no-op for this function's lifetime, and whatever it was is restored.
  was <- options(browser = function(url) invisible(url))
  on.exit(options(was), add = TRUE)

  qmds <- unlist(lapply(dirs, function(d)
    list.files(d, pattern = "[.]qmd$", recursive = TRUE, full.names = TRUE)))
  # The same rule the other guards apply: `_book/` is output, and a
  # `_`-prefixed file is one the author has deliberately withheld.
  qmds <- qmds[!grepl("/_", qmds, fixed = TRUE)]

  # The shared fixtures every chapter includes. Read as a chapter would read it,
  # so the tables here are the tables the examples ran against.
  shared <- new.env(parent = globalenv())
  setup <- file.path(dirs[[1]], "_setup.qmd")
  if (file.exists(setup)) {
    for (code in r_chunks(readLines(setup, warn = FALSE))) {
      try(eval(parse(text = paste(code$code, collapse = "\n")), envir = shared),
          silent = TRUE)
    }
  }

  # **Asked the other way round: which errors mean this guard is broken?**
  # Listing what a refusal looks like was tried first and was wrong, because
  # refusals come from three places with three shapes: the engine's diagnostics,
  # the bindings' own messages, and the one refusal that fires from inside the
  # query and arrives wrapped by the driver. Any positive list misses one of
  # them and reports a working refusal as a fault.
  #
  # What is short and stable is the other list: the errors R raises when this
  # file has failed to give a chunk what it needs.
  harness <- "object '[^']*' not found|could not find function|unused argument"

  checked <- 0L
  quiet <- character()
  wrong <- character()

  for (f in qmds) {
    lines <- readLines(f, warn = FALSE)
    chunks <- r_chunks(lines)
    if (!any(vapply(chunks, function(c) c$tolerates, logical(1)))) next

    # Chapter-local tables are defined by earlier chunks, so the chapter is
    # walked in order and every chunk before the last refusal is run for its
    # side effects. An ordinary chunk that fails here is left alone: the render
    # is what covers those, and stopping would hide the refusals after it.
    # **The fixtures are copied in rather than inherited**, and that is not a
    # style choice. A verb looks a table up with `mget`, which does not search
    # enclosing environments, so a chapter env whose *parent* held `products`
    # reported that there was no such table. The refusal that came back was a
    # real one for the wrong reason, which is the failure mode this guard exists
    # to avoid in the book and must not have itself.
    chapter <- new.env(parent = globalenv())
    for (found in ls(shared)) assign(found, get(found, envir = shared), envir = chapter)
    last <- max(which(vapply(chunks, function(c) c$tolerates, logical(1))))

    # Chunks are evaluated from the file's own directory, because that is
    # knitr's rule and a chapter is allowed to read a file by a path relative
    # to itself. Evaluating from the repository root quietly broke the first
    # chapter that did. The restore is immediate rather than `on.exit`,
    # because the next file's path is relative to where this loop stands.
    owd <- getwd()
    on.exit(setwd(owd), add = TRUE)
    setwd(dirname(f))

    for (i in seq_len(last)) {
      chunk <- chunks[[i]]
      code <- chunk$code[nzchar(trimws(chunk$code))]
      if (!length(code)) next
      where <- sprintf("%s:%d", sub("^.*book/", "", f), chunk$line)

      if (!chunk$tolerates) {
        # Run for its tables, not for its output. A chapter prints queries and
        # translations, and none of that belongs in this guard's report.
        try(invisible(capture.output(
          invisible(capture.output(
            eval(parse(text = paste(code, collapse = "\n")), envir = chapter),
            type = "message"
          ))
        )), silent = TRUE)
        next
      }

      checked <- checked + 1L
      # An assumption is reported on stderr and is not a refusal, so it is
      # caught here rather than printed into this guard's own report.
      outcome <- tryCatch({
        value <- NULL
        invisible(capture.output(
          value <- eval(parse(text = paste(code, collapse = "\n")), envir = chapter),
          type = "message"
        ))
        # **Nothing has run yet if this is a pipeline.** Printing is what forces
        # one in a chapter, so the check forces one here, or a refusal that no
        # longer refuses would look exactly like a refusal that does.
        if (inherits(value, "god_pipeline")) {
          invisible(capture.output(collect(value), type = "message"))
          "the pipeline ran and returned a table"
        } else {
          "it evaluated without complaint"
        }
      }, error = function(e) conditionMessage(e))

      if (!grepl("^(the pipeline ran|it evaluated)", outcome)) {
        if (grepl(harness, outcome)) {
          wrong <- c(wrong, sprintf("%s: %s", where,
                                    trimws(strsplit(outcome, "\n")[[1]][1])))
        }
        next
      }
      quiet <- c(quiet, sprintf("%s (%s): %s", where, outcome,
                                paste(trimws(code), collapse = " ")))
    }

    setwd(owd)
  }

  if (!checked) {
    stop("check_refusals: found no `error: true` chunks, so the scan is broken ",
         "rather than the book", call. = FALSE)
  }

  if (length(quiet) || length(wrong)) {
    if (length(quiet)) {
      cat("FAIL: shown as refusals, and they did not refuse\n")
      for (line in quiet) cat("  ", line, "\n", sep = "")
      cat("  Either the grammar stopped refusing, or the prose should stop saying it does.\n")
    }
    if (length(wrong)) {
      cat("FAIL: these failed for a reason that is not a refusal\n")
      for (line in wrong) cat("  ", line, "\n", sep = "")
      cat("  That is this guard missing a table, not the book being wrong. Fix the guard.\n")
    }
    stop(sprintf("check_refusals: %d chunk(s) wrong", length(quiet) + length(wrong)),
         call. = FALSE)
  }

  unmarked <- unmarked_python_refusals(qmds)
  if (length(unmarked)) {
    cat("FAIL: Python refusal chunks with no `#| classes: refusal`\n")
    for (line in unmarked) cat("  ", line, "\n", sep = "")
    cat("  Without it the twin renders as an ordinary result while the R tab is\n")
    cat("  shaded as a refusal, so the two tabs of one sentence disagree on the page.\n")
    stop(sprintf("check_refusals: %d unmarked Python refusal(s)", length(unmarked)),
         call. = FALSE)
  }

  cat("PASS: every documented refusal refuses (", checked, "chunks )\n")
  cat("PASS: every Python refusal is marked for the stylesheet\n")
  invisible(TRUE)
}

# The R half of a refusal raises, so Quarto marks its output `cell-output-error`
# and the stylesheet can find it. The Python half catches and prints, which lands
# in `cell-output-stdout` beside every ordinary table in the book. So the Python
# chunk carries `#| classes: refusal`, which Quarto puts on the cell, and the
# stylesheet reaches the output through it.
#
# **Nothing else can catch a missing one.** The chunk still runs, the message
# still prints, the render still exits 0, and the only symptom is a tab that is
# not shaded on a page where its partner is. That is invisible to every other
# guard here, all of which ask whether something *ran*.
unmarked_python_refusals <- function(qmds) {
  out <- character()
  for (f in qmds) {
    lines <- readLines(f, warn = FALSE)
    open <- grep("^\\s*```\\{python\\}\\s*$", lines)
    for (start in open) {
      close <- grep("^\\s*```\\s*$", lines)
      close <- close[close > start]
      if (!length(close)) next
      body <- lines[(start + 1L):(close[1] - 1L)]
      if (!any(grepl("except GodError", body))) next
      if (any(grepl("^\\s*#\\|\\s*classes:.*\\brefusal\\b", body))) next
      out <- c(out, sprintf("%s:%d", sub("^.*book/", "", f), start))
    }
  }
  out
}

# Every `{r}` chunk in a file, with its line, its code, and whether it is marked
# to tolerate an error.
r_chunks <- function(lines) {
  out <- list()
  i <- 1L
  while (i <= length(lines)) {
    if (grepl("^\\s*```\\{r\\}\\s*$", lines[i])) {
      start <- i
      i <- i + 1L
      body <- character()
      while (i <= length(lines) && !grepl("^\\s*```\\s*$", lines[i])) {
        body <- c(body, lines[i])
        i <- i + 1L
      }
      options <- grep("^\\s*#\\|", body, value = TRUE)
      out[[length(out) + 1L]] <- list(
        line = start,
        code = body[!grepl("^\\s*#\\|", body)],
        tolerates = any(grepl("error:\\s*true", options))
      )
    }
    i <- i + 1L
  }
  out
}

if (sys.nframe() == 0L) check_refusals()

File: book/check_refusals.py
"""check_refusals.py — the Python half of every documented refusal.

Run from the repository root; run by the Python test suite.

**This is `check_refusals.R`'s twin, and the tabsets are why it exists.** Every
refusal in the book is shown twice, once per language. The R half is a chunk
marked `#| error: true`, and the R guard forces each one and asserts it still
refuses. The Python half is a `try/except GodError` block, which is not marked
at all: it prints the refusal when the grammar refuses, and it would print a
table just as happily the day the grammar stopped. Two languages proving each
other is the whole point of showing both, and until this file only one of them
was proved.

**A pipeline is lazy here too.** `sales >> keep(...)` builds a sentence and
raises nothing; `collect`, or printing, is what hands it to the engine. Every
refusal block in the book collects inside the `try`, and where one does not,
the pipeline it builds is forced here, or a refusal that no longer refuses
would look exactly like one that does.

**An error is not automatically a refusal**, so the same rule as the R guard:
the short, stable list is the errors that mean *this file* failed to give a
chunk what it needs (a missing name, a missing import). Anything else that
raises is the grammar, or the driver carrying the grammar's answer, and both
of those are refusals.
"""

import ast
import contextlib
import io
import os
import sys
from contextlib import redirect_stderr, redirect_stdout
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
BOOK = ROOT / "book"


@contextlib.contextmanager
def working_directory(directory: Path):
    """Run a chapter's chunks from the chapter's own directory, as knitr does.

    **This guard runs from the repository root and the render does not**, which
    did not matter until a chapter's table came from `god_table`. That walks up
    from the working directory for a `data/<name>.csv`, so from `book/chapters/`
    it finds `book/data/` and from the root it finds nothing and reaches for the
    published copy — a table not yet published is then a `NameError` here and a
    rendered page there. Matching knitr is the fix rather than special-casing
    the walk: a guard that executes chunks should execute them where they run.
    """
    was = Path.cwd()
    os.chdir(directory)
    try:
        yield
    finally:
        os.chdir(was)


def python_chunks(lines: list[str]) -> list[dict]:
    """Every ```{python} chunk, with its first line number and its code."""
    out = []
    i = 0
    while i < len(lines):
        if lines[i].strip() == "```{python}":
            start = i + 1
            i += 1
            body = []
            while i < len(lines) and lines[i].strip() != "```":
                body.append(lines[i])
                i += 1
            out.append({"line": start, "code": "\n".join(body)})
        i += 1
    return out


def catches_goderror(node: ast.Try) -> bool:
    for handler in node.handlers:
        kind = handler.type
        names = []
        if isinstance(kind, ast.Name):
            names = [kind.id]
        elif isinstance(kind, ast.Tuple):
            names = [e.id for e in kind.elts if isinstance(e, ast.Name)]
        if "GodError" in names:
            return True
    return False


def run_statements(statements, namespace, filename):
    """Execute statements; the last one, if an expression, hands back its value."""
    value = None
    for i, statement in enumerate(statements):
        module = ast.Module(body=[statement], type_ignores=[])
        ast.fix_missing_locations(module)
        if i == len(statements) - 1 and isinstance(statement, ast.Expr):
            expression = ast.Expression(body=statement.value)
            ast.fix_missing_locations(expression)
            value = eval(compile(expression, filename, "eval"), namespace)
        else:
            exec(compile(module, filename, "exec"), namespace)
    return value


def force(value, namespace):
    """A pipeline that was built and never run is run now.

    Printing is what forces one in a chapter, so the check forces one here.
    """
    if value is not None and hasattr(value, "written"):
        namespace["collect"](value)
        return "the pipeline ran and returned a table"
    return "it evaluated without complaint"


def main() -> int:
    if not BOOK.exists():
        print("SKIP: no book/ here, so there are no refusal chunks to check")
        return 0

    # The shared fixtures, read the way a chapter reads them.
    shared: dict = {}
    setup = BOOK / "_setup.qmd"
    if setup.exists():
        for chunk in python_chunks(setup.read_text().splitlines()):
            try:
                exec(compile(chunk["code"], str(setup), "exec"), shared)
            except Exception:
                pass

    # The errors that mean this guard is broken rather than the book: Python's
    # spellings of "object not found" and "could not find function".
    harness = (NameError, ImportError)

    checked = 0
    quiet: list[str] = []
    wrong: list[str] = []

    qmds = sorted(p for p in BOOK.rglob("*.qmd") if "/_" not in str(p))
    for f in qmds:
        lines = f.read_text().splitlines()
        chunks = python_chunks(lines)
        parsed = []
        for chunk in chunks:
            try:
                tree = ast.parse(chunk["code"])
            except SyntaxError:
                tree = None
            parsed.append(tree)
        refusing = [
            any(isinstance(n, ast.Try) and catches_goderror(n) for n in ast.walk(t))
            if t is not None
            else False
            for t in parsed
        ]
        if not any(refusing):
            continue

        # Chapter-local tables come from earlier chunks, so the chapter is
        # walked in order up to its last refusal. The fixtures are copied in,
        # not inherited, for the R guard's reason: a chunk has to see exactly
        # the tables a chapter gave it.
        namespace = dict(shared)
        last = max(i for i, r in enumerate(refusing) if r)

        with working_directory(f.parent):
            for i in range(last + 1):
                chunk, tree = chunks[i], parsed[i]
                where = f"{f.relative_to(ROOT)}:{chunk['line']}"
                if tree is None:
                    continue

                if not refusing[i]:
                    # Run for its tables, not for its output.
                    try:
                        with redirect_stdout(io.StringIO()), redirect_stderr(io.StringIO()):
                            exec(compile(tree, str(f), "exec"), namespace)
                    except Exception:
                        pass
                    continue

                # A refusal chunk can carry setup around its `try`, so the chunk
                # is executed statement by statement and only the `try` bodies
                # are held to the promise.
                for node in tree.body:
                    if isinstance(node, ast.Try) and catches_goderror(node):
                        checked += 1
                        shown = " ".join(
                            ast.get_source_segment(chunk["code"], node.body[0]).split()
                        ) if node.body else ""
                        try:
                            with redirect_stdout(io.StringIO()), redirect_stderr(io.StringIO()):
                                value = run_statements(node.body, namespace, str(f))
                                outcome = force(value, namespace)
                        except harness as e:
                            wrong.append(f"{where}: {type(e).__name__}: {e}")
                            continue
                        except Exception:
                            continue  # it refused, which is the promise kept
                        quiet.append(f"{where} ({outcome}): {shown}")
                    else:
                        try:
                            with redirect_stdout(io.StringIO()), redirect_stderr(io.StringIO()):
                                run_statements([node], namespace, str(f))
                        except Exception:
                            pass

    if not checked:
        print(
            "check_refusals.py: found no try/except GodError chunks, "
            "so the scan is broken rather than the book"
        )
        return 1

    if quiet or wrong:
        if quiet:
            print("FAIL: shown as refusals, and they did not refuse")
            for line in quiet:
                print(f"  {line}")
            print("  Either the grammar stopped refusing, or the prose should stop saying it does.")
        if wrong:
            print("FAIL: these failed for a reason that is not a refusal")
            for line in wrong:
                print(f"  {line}")
            print("  That is this guard missing a table, not the book being wrong. Fix the guard.")
        return 1

    print(f"PASS: every Python refusal refuses ( {checked} chunks )")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
